src/builder/commands/ubuntu.rs

```diff
@@ -626,6 +645,46 @@ impl Distro {
             Ok(())
         }).map_err(|x| io::Error::new(io::ErrorKind::Other, x)).and_then(|x| x)
     }
+
+    #[cfg(feature="containers")]
+    fn copy_apt_archives_to_cache(&self, overlay_dir: &Path, cache_dir: &Path) -> io::Result<()> {
+        use rand::{thread_rng, Rng};
+        use rand::distributions::Alphanumeric;
+
+        let cache_dir = cache_dir.join("archives");
+        Dir::new(&cache_dir).create()?;
+
+        let archives_dir = overlay_dir.join("upper/archives");
+        if !archives_dir.exists() {
+            return Ok(());
+        }
+        ScanDir::files().read(&archives_dir, |iter| {
+            for (entry, name) in iter {
+                if !name.ends_with(".deb") { continue }
+                if !entry.file_type()?.is_file() { continue }
+                let src = entry.path();
+                let tmp_prefix: String = thread_rng()
+                    .sample_iter(&Alphanumeric)
+                    .take(6)
+                    .map(char::from)
+                    .collect();
+                let tmp = cache_dir.join(format!(".{}-{}.part", tmp_prefix, name));
+                let dst = cache_dir.join(name);
+                if !dst.exists() {
+                    // Cannot rename without copying due to "Cross-device link (os error 18)"
+                    copy(&src, &tmp)?;
+                    copy_utime(&src, &tmp)?;
+                    rename(tmp, dst)?;
+                }
+            }
+            Ok(())
```

Comment:
Also we should cleanup the cache from the container i.e. `remove_dir_all(overlay_dir.join("tmp_cache"))` (that leaves only dirs we we can `rename` to the cache dir, but that's also nice to remove).

Reply:
Could you explain what do you want?

Reply:
You create directory `.vagga/.roots/.tmp.container.deadbee/upper` (or `x/cache` or `x/tmp_cache` or whatever) which contains cache, temporarily. Then you copy all the data into `.vagga/.cache` for persistent storage, and the original "upper" directory is left besides the container in `.vagga/.roots/container.deadbee/upper`, while it's useless right after build.

Reply:
Ah, I see it's done now in `guard.rs`